File: fab-core/src/main/java/eu/stratosphere/fab/core/beans/experiment/Experiment.java
package eu.stratosphere.fab.core.beans.experiment;

import java.util.*;

import eu.stratosphere.fab.core.beans.ExecutionContext;
import eu.stratosphere.fab.core.beans.system.ExperimentRunner;
import eu.stratosphere.fab.core.beans.system.JavaExperimentRunner;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public class Experiment {

    static Logger logger = LoggerFactory.getLogger(Experiment.class);

    public final String name;

    public final ExperimentRunner runner;

    public final Map<String, String> options;

    public final List<String> arguments;

    public Experiment(ExperimentRunner runner) {
        this.name = "Experiment " + UUID.randomUUID().toString();
        this.runner = runner;
        this.options = new HashMap<String, String>();
        this.arguments = new ArrayList<String>();
    }

    public Experiment(String name, ExperimentRunner runner) {
        this.name = name;
        this.runner = runner;
        this.options = new HashMap<String, String>();
        this.arguments = new ArrayList<String>();
    }

    public Experiment(String name, ExperimentRunner runner, Map<String, String> options) {
        this.name = name;
        this.runner = runner;
        this.options = options;
        this.arguments = new ArrayList<String>();
    }

    public Experiment(String name, ExperimentRunner runner, List<String> arguments) {
        this.name = name;
        this.runner = runner;
        this.options = new HashMap<String, String>();
        this.arguments = arguments;
    }

    public Experiment(String name, ExperimentRunner runner, Map<String, String> options, List<String> arguments) {
        this.name = name;
        this.runner = runner;
        this.options = options;
        this.arguments = arguments;
    }

    public void run(ExecutionContext context) {
        logger.info("Running experiment XXX with runner " + runner.name);
        //this.runner.run(context);
    }
}

File: fab-core/src/main/scala/eu/stratosphere/fab/core/beans/experiment/Experiment.scala
package eu.stratosphere.fab.core.beans.experiment

import eu.stratosphere.fab.core.beans.system.{FileSystem, Lifespan, ExperimentRunner, System}
import eu.stratosphere.fab.core.{ExecutionContext, Node}
import org.slf4j.LoggerFactory
import com.typesafe.config.ConfigFactory
import java.io.{FileNotFoundException, File}
import scala.collection.JavaConverters._

/**
 * Created by felix on 12.06.14.
 */
class Experiment (val runner: ExperimentRunner,
                  val arguments: List[String] = List())
                  extends Node {

  final val logger = LoggerFactory.getLogger(this.getClass)
  final val ExpConf = ConfigFactory.load(arguments(0))
  final val config = ConfigFactory.load()
  final val sequence = ExpConf.getIntList("experiment.runs").asScala.toList


  def run(ctx: ExecutionContext) = {
    logger.info("Starting experiment sequence with %d element(s)...".format(sequence.length))
    for (num <- 0 to sequence.length - 1) { // start experiment sequence
      logger.info("Starting element %d/%d with %d repetitions...".format(num+1, sequence.length, sequence(num).toInt))
      val job: String = arguments(1)
      val output = new File(config.getString("paths.hadoop.v1.output"), "output.txt")
      val results = config.getString("paths.hadoop.v1.results")

      val fs: List[FileSystem] = getFileSystems(ctx)
      def copyInput: List[File] = fs map { x: FileSystem => x.setInput(new File(arguments(2)))}
      def copyResults(run: Int) = fs map { x: FileSystem => x.getOutput(output, new File(results, "results_SeqRun_%d_run_%d".format(num+1, run)))}

      //TODO check arguments for correctness

      for (run <- 1 to sequence(num)) { // repeat experiment with sequence number num
        logger.info("Running repetition %d/%d".format(run, sequence(num).toInt))
        setUpComponents(ctx.expGraph.reverse.directDependencies(runner)) // setup runs on reversed graph
        runner.run(job, copyInput, output)
        copyResults(run)
        tearDownComponents(ctx.expGraph.directDependencies(runner))
      }
      //TODO find good solution to save results of a run
    }
  }

  def setUpComponents(systems: List[Node]) = {
    for(s <- systems ) yield {
      s match {
      case s: System => if (s.lifespan == Lifespan.EXPERIMENT) s.setUp()
      case x => x
      }
    }
  }

  def tearDownComponents(systems: List[Node]) = {
    for(s <- systems ) yield {
      s match {
        case s: System => if (s.lifespan == Lifespan.EXPERIMENT) s.tearDown()
        case x => x
      }
    }
  }

  //TODO make filesystem a component of experiment?
  def getFileSystems(ctx: ExecutionContext): List[FileSystem] = {
    ctx.expGraph.directDependencies(this) flatMap {
      case fs: FileSystem => Some(fs)
      case _ => None
    }
  }

  def update() = {
    // TODO find components with lifespan experiment sequence
    // update all of their dependend components but don't update them twice
    // => build update graph for every exp_seq component, join graphs
  }


}
